### PR TITLE
feat: local registry

### PR DESCRIPTION
**Pull request by GustavBaumgart**
## Description

A local registry saves the metrics (which are currently implemented on a per-round basis) to a ~/flame-log directory. The structure of this directory is flame-log --> job_id --> task_id --> {model, params, metrics}. Each metric is saved in its own separate csv.

Additionally, the method signatures for __call__ were changed, and the method signatures for save_params were aligned to what the MLflow registry has used.

## Type of Change

- [ ] Bug Fix
- [x] New Feature
- [ ] Breaking Change
- [ ] Refactor
- [ ] Documentation
- [ ] Other (please describe)

## Checklist

- [ ] I have read the [contributing guidelines](/CONTRIBUTING.md)
- [ ] Existing issues have been referenced (where applicable)
- [ ] I have verified this change is not present in other open pull requests
- [ ] Functionality is documented
- [ ] All code style checks pass
- [ ] New code contribution is covered by automated tests
- [ ] All new and existing tests pass

### CODE REVIEW

File: lib/python/flame/config.py [hunk 1]
```diff
@@ -48,6 +48,7 @@ class RegistryType(str, Enum):
 
     DUMMY = "dummy"
     MLFLOW = "mlflow"
+    LOCAL = "local"
```
Comment (myungjin) [inline lib/python/flame/config.py:51]: please order them in alphanumeric order.

File: lib/python/flame/mode/distributed/trainer.py [hunk 2]
```diff
@@ -70,7 +70,7 @@ def internal_init(self) -> None:
         """Initialize internal state for role."""
         self.registry_client = registry_provider.get(self.config.registry.sort)
         # initialize registry client
-        self.registry_client(self.config.registry.uri, self.config.job.job_id)
+        self.registry_client(self.config.registry.uri, self.config)
```
Comment (myungjin) [inline lib/python/flame/mode/distributed/trainer.py:73]: Passing `self.config` alone may be sufficient since registry.uri can be accessed from config.

File: lib/python/flame/mode/horizontal/syncfl/top_aggregator.py [hunk 3]
```diff
@@ -81,7 +81,7 @@ def internal_init(self) -> None:
 
         self.registry_client = registry_provider.get(self.config.registry.sort)
         # initialize registry client
-        self.registry_client(self.config.registry.uri, self.config.job.job_id)
+        self.registry_client(self.config.registry.uri, self.config)
```
Comment (myungjin) [inline lib/python/flame/mode/horizontal/syncfl/top_aggregator.py:84]: same here

File: lib/python/flame/mode/horizontal/syncfl/trainer.py [hunk 4]
```diff
@@ -71,7 +71,7 @@ def internal_init(self) -> None:
 
         self.registry_client = registry_provider.get(self.config.registry.sort)
         # initialize registry client
-        self.registry_client(self.config.registry.uri, self.config.job.job_id)
+        self.registry_client(self.config.registry.uri, self.config)
```
Comment (myungjin) [inline lib/python/flame/mode/horizontal/syncfl/trainer.py:74]: same here

File: lib/python/flame/registries.py [hunk 5]
```diff
@@ -21,6 +21,7 @@
 from .object_factory import ObjectFactory
 from .registry.dummy import DummyRegistryClient
 from .registry.mlflow import MLflowRegistryClient
```
Comment (myungjin) [inline lib/python/flame/registries.py:23]: please refactor other imports to use absolute path

File: lib/python/flame/registry/abstract.py [hunk 6]
```diff
@@ -17,13 +17,14 @@
 
 from abc import ABC, abstractmethod
 from typing import Any, Optional
+from ..config import (Hyperparameters, Config)
```
Comment (myungjin) [inline lib/python/flame/registry/abstract.py:20]: please use absolute path

File: lib/python/flame/registry/dummy.py [hunk 7]
```diff
@@ -17,13 +17,14 @@
 
 from typing import Any, Optional
 
+from ..config import (Hyperparameters, Config)
 from .abstract import AbstractRegistryClient
```
Comment (myungjin) [inline lib/python/flame/registry/dummy.py:21]: refactor to use absolute path

File: lib/python/flame/registry/local.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+"""Local registry client."""
+
+from typing import Any, Optional
+from .abstract import AbstractRegistryClient
+
+from ..config import Hyperparameters
+from ..config import Config
```
Comment (myungjin) [inline lib/python/flame/registry/local.py:22]: please use absolute path

File: lib/python/flame/registry/local.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+import logging
+import os
+import csv
+import pickle
+from collections import defaultdict
+from flame.common.util import (get_ml_framework_in_use, MLFramework)
```
Comment (myungjin) [inline lib/python/flame/registry/local.py:28]: Please put a structure on the imports.

Please take a look at the following example: https://github.com/cisco-open/flame/blob/main/lib/python/flame/mode/horizontal/syncfl/top_aggregator.py#L18-L44

There are a couple of python auto-formatter: autopep8. Please consider to use one.

File: lib/python/flame/registry/local.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    
+    def __call__(self, uri: str, config: Config) -> None:
+        """Initialize the instance."""
+        self.job_id = config.job.job_id
+        self.task_id = config.task_id
+
+    def setup_run(self, name: str) -> None:
+        """Set up a run."""
+        # set up directories
+        home_dir = os.path.expanduser("~")
```
Comment (myungjin) [inline lib/python/flame/registry/local.py:52]: Perhaps it may be better to use `pathlib`?
An example usage:
```
from pathlib import Path
print(Path.home())
```

File: lib/python/flame/registry/local.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        log_dir = os.path.join(home_dir, "flame-log")
+        self.registry_path = os.path.join(log_dir, self.job_id, self.task_id)
+        
+        os.system(f'rm -rf {self.registry_path}')
```
Comment (myungjin) [inline lib/python/flame/registry/local.py:56]: perhaps better to use `shutil.rmtree()`?

File: lib/python/flame/registry/mlflow.py [hunk 12]
```diff
@@ -23,7 +23,7 @@
 from mlflow.exceptions import MlflowException
 from mlflow.protos.databricks_pb2 import RESOURCE_ALREADY_EXISTS, ErrorCode
 
-from ..config import Hyperparameters
+from ..config import (Hyperparameters, Config)
 from .abstract import AbstractRegistryClient
```
Comment (myungjin) [inline lib/python/flame/registry/mlflow.py:27]: please use absolute path

File: lib/python/flame/registry/local.py [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+        os.makedirs(self.registry_path)
+        os.mkdir(os.path.join(self.registry_path, "metrics"))
+        os.mkdir(os.path.join(self.registry_path, "model"))
+        os.mkdir(os.path.join(self.registry_path, "params"))
+        #os.mkdir(os.path.join(self.registry_path, "artifact"))
```
Comment (myungjin) [inline lib/python/flame/registry/local.py:61]: remove it if this is not needed

File: lib/python/flame/registry/local.py [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+        # set up directories
+        home_dir = os.path.expanduser("~")
+        log_dir = os.path.join(home_dir, "flame-log")
+        self.registry_path = os.path.join(log_dir, self.job_id, self.task_id)
+        
+        os.system(f'rm -rf {self.registry_path}')
+        os.makedirs(self.registry_path)
+        os.mkdir(os.path.join(self.registry_path, "metrics"))
+        os.mkdir(os.path.join(self.registry_path, "model"))
+        os.mkdir(os.path.join(self.registry_path, "params"))
```
Comment (myungjin) [inline lib/python/flame/registry/local.py:60]: Instead of lines 57-60, how about the following?
```
for dir in ["metrics", "model", "params"]:
    os.makedirs(os.path.join(self.registry_path, dir))
```

**Review by myungjin — COMMENTED** [review]
left comments.

File: lib/python/flame/registry/local.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+        #os.mkdir(os.path.join(self.registry_path, "artifact"))
+        
+        # version tracking
+        self.param_version = 1
+        self.model_versions = defaultdict(lambda: 1)
+
+    def save_metrics(self, epoch: int, metrics: Optional[dict[str,
+                                                              float]]) -> None:
```
Comment (jaemin-shin) [inline lib/python/flame/registry/local.py:68]: better merge line 68 with 67. Setting up formatter would help

File: lib/python/flame/registry/local.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+    def cleanup(self) -> None:
+        """Clean up resources."""
+        pass
+
+    def save_model(self, name: str, model: Any) -> None:
+        """Save a model in a model registry."""
+        model_folder = os.path.join(self.registry_path, "model", name)
+        if not os.path.exists(model_folder):
+            os.mkdir(model_folder)
```
Comment (jaemin-shin) [inline lib/python/flame/registry/local.py:96]:
```suggestion
            os.makedirs(model_folder, exist_ok=True)
```

You may consider merging 95 and 96 into one line like this

File: lib/python/flame/registry/local.py [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+        model_path = os.path.join(self.registry_path, "model", name, str(version))
+        if ml_framework == MLFramework.PYTORCH:
+            # is not safe but matches behavior of MLFlow registry
+            import torch
+            return torch.load(model_path)
+        elif ml_framework == MLFramework.TENSORFLOW:
+            import tensorflow
+            return tensorflow.keras.models.load_model(model_path)
+        
+        return None
```
Comment (jaemin-shin) [inline lib/python/flame/registry/local.py:124]: how about we raise an error like ModuleNotFoundError, instead of return None?

**Review by jaemin-shin — COMMENTED** [review]
Left some comments. Agree to Myungjin's comments too!

File: lib/python/flame/registry/local.py [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+        """Set up a run."""
+        # set up directories
+        home_dir = Path.home()
+        log_dir = os.path.join(home_dir, "flame-log")
+        self.registry_path = os.path.join(log_dir, self.job_id, self.task_id)
+
+        if os.path.exists(self.registry_path):
+            shutil.rmtree(self.registry_path)
+
+        os.makedirs(self.registry_path)
```
Comment (myungjin) [inline lib/python/flame/registry/local.py:63]: This doesn't seem necessary because line 65 will create any intermediate directories in the path.

File: lib/python/flame/registry/local.py [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+        for directory in ["metrics", "model", "params"]:
```
Comment (myungjin) [inline lib/python/flame/registry/local.py:64]: These ("metrics", "model" and "params") are used in several places in the code.
In this case, making them as constants may be better to reduce the chance of introducing a bug.

File: lib/python/flame/registry/local.py [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+            filename = os.path.join(self.registry_path, "metrics", metric)
+            if os.path.exists(filename):
+                with open(filename, "a") as file:
+                    csv_writer = csv.writer(file)
+                    csv_writer.writerow([epoch, metrics[metric]])
+            else:
+                with open(filename, "w") as file:
+                    csv_writer = csv.writer(file)
+                    csv_writer.writerow(["round", metric])
+                    csv_writer.writerow([epoch, metrics[metric]])
```
Comment (myungjin) [inline lib/python/flame/registry/local.py:83]: I think lines 74-83 can be simplified in the following way:
```
exists = os.path.exists(filename):
with open(filename, "a+") as file:
    csv_writer = csv.writer(file)
    csv_writer.writerow(["round", metric]) if not exists else None
    csv_writer.writerow([epoch, metrics[metric]])
```

**Review by myungjin — COMMENTED** [review]
left a few comments; please consider to address them

**Review by myungjin — COMMENTED** [review]
left a few comments; please consider to address them

File: lib/python/flame/registry/local.py [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+        """Save a model in a model registry."""
+        model_folder = os.path.join(self.registry_path, "model", name)
+        os.makedirs(model_folder, exist_ok=True)
+
+        ml_framework = get_ml_framework_in_use()
+
+        if ml_framework == MLFramework.PYTORCH:
+            import torch
+
+            # is not safe but matches behavior of MLFlow registry
```
Comment (jaemin-shin) [inline lib/python/flame/registry/local.py:107]: what part of the code does this comment indicates? could be confusing as there are same comment at different location in line 124. Also, the comment itself could be more detailed to be helpful, by explaining which behavior of MLFlow registry it refers to

**Review by jaemin-shin — COMMENTED** [review]
LGTM for me! Left one minor comment.

**Review by myungjin — APPROVED** [review]
lgtm
